internal/local/structure.go

```diff
@@ -0,0 +1,222 @@
+	Services      map[string]map[string]string `json:"services,omitempty"`
+}
+
+// ServiceStructure represents the Realm app service structure
+type ServiceStructure struct {
+	Config           map[string]interface{}   `json:"config,omitempty"`
+	IncomingWebhooks []map[string]interface{} `json:"incoming_webhooks"`
+	Rules            []map[string]interface{} `json:"rules"`
+}
+
+func parseEnvironments(rootDir string) (map[string]map[string]interface{}, error) {
+	out := map[string]map[string]interface{}{}
+
+	dw := directoryWalker{
+		path:      filepath.Join(rootDir, NameEnvironments),
+		onlyFiles: true,
+	}
+	if err := dw.walk(func(file os.FileInfo, path string) error {
+		o, err := parseJSON(path)
+		if err != nil {
+			return err
+		}
+		out[file.Name()] = o
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+
+	if len(out) == 0 {
+		return nil, nil
+	}
+	return out, nil
+}
+
+func parseFunctions(rootDir string) ([]map[string]interface{}, error) {
+	if ok, err := fileExists(rootDir); err != nil {
+		return nil, err
+	} else if !ok {
+		return nil, nil
+	}
+
+	var out []map[string]interface{}
+
+	dw := directoryWalker{path: rootDir, onlyDirs: true}
+	if walkErr := dw.walk(func(file os.FileInfo, path string) error {
+		if strings.Contains(path, nameNodeModules) {
+			return nil // skip node_modules
+		}
+
+		config, configErr := parseJSONConfig(path)
+		if configErr != nil {
+			return configErr
+		}
+
+		src, srcErr := parseJavascriptSource(path)
+		if srcErr != nil {
+			return srcErr
+		}
+
+		out = append(out, map[string]interface{}{
+			NameConfig: config,
+			NameSource: src,
+		})
+		return nil
+	}); walkErr != nil {
+		return nil, walkErr
+	}
+
+	return out, nil
+}
+
+func parseGraphQL(rootDir string) (GraphQLStructure, bool, error) {
+	dir := filepath.Join(rootDir, NameGraphQL)
+
+	if ok, err := fileExists(dir); err != nil {
+		return GraphQLStructure{}, false, err
+	} else if !ok {
+		return GraphQLStructure{}, false, nil
+	}
+
+	config, configErr := parseJSONConfig(dir)
+	if configErr != nil {
+		return GraphQLStructure{}, false, configErr
+	}
+
+	customResolvers, customResolversErr := parseJSONFiles(filepath.Join(dir, NameCustomResolvers))
+	if customResolversErr != nil {
+		return GraphQLStructure{}, false, customResolversErr
+	}
+
+	return GraphQLStructure{config, customResolvers}, true, nil
+}
+
+func parseJavascript(rootDir string, file File) (string, error) {
+	src, err := ioutil.ReadFile(filepath.Join(rootDir, file.String()))
+	if err != nil {
+		return "", err
+	}
+	return string(src), nil
+}
+
+func parseJavascriptSource(rootDir string) (string, error) {
+	return parseJavascript(rootDir, FileSource)
+}
+
+func parseJSON(path string) (map[string]interface{}, error) {
+	data, dataErr := readFile(path)
+	if dataErr != nil {
+		return nil, dataErr
+	}
+
+	var out map[string]interface{}
+	if err := unmarshalJSON(data, &out); err != nil {
+		return nil, err
+	}
+	return out, nil
+}
```

Comment:
do we really need this func? doesn't seem like we gain too much other than abstracting away 9 lines of code 🤷🏻

Reply:
you *really* don't want me to have this helper (it's been removed and added back now about 2-3 different times in various forms).

we use this function/chunk of code currently 24 times in our app structure/parsing code, and that's not yet even our cap.  it really does have an impact on the code heft/readability in this case

Reply:
😅  alright I'll stop commenting on this. carry on...